src/EnergyPlus/InternalHeatGains.cc

```diff
@@ -8054,31 +8053,24 @@ namespace InternalHeatGains {
 
         // PURPOSE OF THIS SUBROUTINE:
         // utility to retrieve index pointer to a specific internal gain
+        // the subroutine returns the index of matched internal gain device or -1 if no match found.
 
         // SUBROUTINE LOCAL VARIABLE DECLARATIONS:
-        bool Found;
         int DeviceNum;
-
-        Found = false;
-
         if (state.dataHeatBal->ZoneIntGain(ZoneNum).NumberOfDevices == 0) {
             DeviceIndex = -1;
-            ErrorFound = true;
-            return;
+            return DeviceIndex;
         }
-
         for (DeviceNum = 1; DeviceNum <= state.dataHeatBal->ZoneIntGain(ZoneNum).NumberOfDevices; ++DeviceNum) {
-            if (UtilityRoutines::SameString(state.dataHeatBal->ZoneIntGain(ZoneNum).Device(DeviceNum).CompObjectName, IntGainName)) {
-                if (state.dataHeatBal->ZoneIntGain(ZoneNum).Device(DeviceNum).CompTypeOfNum != IntGainTypeOfNum) {
-                    ErrorFound = true;
-                } else {
-                    ErrorFound = false;
-                }
-                Found = true;
+            if (UtilityRoutines::SameString(state.dataHeatBal->ZoneIntGain(ZoneNum).Device(DeviceNum).CompObjectName, IntGainName) &&
+                state.dataHeatBal->ZoneIntGain(ZoneNum).Device(DeviceNum).CompTypeOfNum == IntGainTypeOfNum) {
                 DeviceIndex = DeviceNum;
                 break;
+            } else {
+                DeviceIndex = -1;
             }
         }
+        return DeviceIndex;
     }
```

Comment:
If you are returning `DeviceIndex`, why does it also need to be a reference parameter?

Reply:
@amirroth  Thank you for your comments! The returned value for `DeviceIndex` was to replace `ErrorFound` in the original function `GetInternalGainDeviceIndex`. In the same time, `DeviceIndex` in this function (`GetInternalGainDeviceIndex`) needs to be passed back to `state.dataRoomAirMod->RoomAirflowNetworkZoneInfo(ZoneNum).Node(RAFNNodeNum).IntGainsDeviceIndices(gainsLoop)` within the function `void GetRoomAirflowNetworkData` in `RoomAirModelManager`. 
What would be a better way of handling this? Any recommendations? Thanks a lot!

Reply:
Well, if a returned index of -1 implies error then, you can use that to infer the `IntGainError` flag at the call site. If there is a single error mode that can be specified using a return value (the two most common ones are -1 for IndexNotFound and NULL for ObjectNotFound or MemoryReferenceNotFound) then this is a a good pattern, especially if the error must be handled in the direct caller. 

I've noticed EnergyPlus often passes Error parameters through multiple levels of functions. This pattern is less good and has been subsumed by exceptions. I assume that at one point we will do a pass that cleans that up that pattern as well, but probably not in the very near future.